packages/loader/container-loader/src/container.ts

```diff
@@ -2000,9 +2000,10 @@ export class Container
 	}
 
 	private get client(): IClient {
+		const loaderOptionsClient = structuredClone(this.options?.client);
 		const client: IClient =
```

Comment:
Cloning does appear to be the correct solution but as implement it means every use will get a unique IClient object.
It is possible that callers are expecting a stable result (to the extent that `this.options?.client` already doesn't change).
Unless we are sure that callers are not expecting stability, we should consider caching.

Reply:
I did some further testing with a custom IClient interface which includes readonly properties. However, it seems like the bug still occurs with the implementation so it seems like the mutation happens at the runtime. Since the bug specifically results from the mutation of client.details.capabilities, I added Object.freeze on this property to prevent mutation at runtime and it seems to resolve the bug as well. Would this be a better alternative to structuredClone?

Reply:
I didn't know about `Object.freeze`. That is almost helpful, but has two problems. (1) it can deny a modification silently (extremely hard to debug) and (2) a second call to get client will fail in certain cases if I understand correctly.
Computing `client` at ctor (with appropriate cloning) still seems like the best solution.
I am not sure what you mean with "custom IClient interface"; shouldn't need to change IClient signature.

Reply:
I think I may have misinterpreted your original comment and tried to replace the return type on the client-getter such that it returns an interface similar to client but with readonly properties.

Looking at the current implementation, I'm not too sure how to replace the client-getter with a readonly property since there seem to be some intentional changes to client properties such as `client.details.environment` on each get client call.

Reply:
I don't know if it is intentional, but something like that is usually problematic. I think we'll have to find the real intended behavior and history. Often there is an oversight regarding the possibility that original sources might change over time. If it is there for testing convenience, then it should probably go away.